File: src/scubaLive/Dive.java
package scubaLive;

public class Dive {
	
	private DiveTable dtable;
	private int bottomt;
	private int ddepth;
	private char spg;
	private char fpg;
	private boolean ssrequired;
	private boolean safe;
	private int diveid;
	public int[] result;

	public Dive(DiveTable divt, SurfaceTable surft, int id) {
		this.dtable = divt;
		bottomt = 0;
		ddepth = 0;
		diveid = id;
		spg = 'a';
		fpg = 'a';
		ssrequired = false;
		safe = true;
		result = null;
	}
	
	public void updateDive(int time, int depth, char startpg) {
		int maxbt = dtable.maxBT(startpg, depth);
		result = new int[]{time, depth, (int) startpg, (int) 'a', 0, 1};
		
		bottomt = time;
		ddepth = depth;
		spg = startpg;
		this.result [2] = startpg;
		
		//If time requested is past the maximum bottom time at depth & spg set return array to carry user recomendations for UI warnings
		if (maxbt < time || (diveid > 1 && startpg == 'a')) {
			safe = false;
			result[0] = maxbt;
			result[1] = dtable.maxDepth(time, startpg);
			result[5] = 0;
			// find new spg if possible
			if(startpg != 'a') {
				if (dtable.maxBT('a', depth) > time) {
					result[2] = dtable.minPG(time, depth);
					result[3] = dtable.diveFPG((char) result[2], depth, time);
					result[4] = dtable.ssTest((char) result[3], depth) ? 1 : 0;
				}
			}
		}
		else {
			fpg = dtable.diveFPG(startpg, depth, time);
			result[3] = fpg;
			ssrequired = dtable.ssTest(fpg, depth);
			result[4] = ssrequired ? 1 : 0;
		} 
		
		
	}
	
	public void updateDive(char startpg) {
		int maxbt = dtable.maxBT(startpg, ddepth);
		result = new int[]{bottomt, ddepth, (int) startpg, (int) 'a', 0, 1};
		
		spg = startpg;
		result [2] = startpg;
		
		//If time requested is past the maximum bottom time at depth & spg set return array to carry user recomendations for UI warnings
		if (maxbt < bottomt || (diveid > 1 && spg == 'a')) {
			safe = false;
			result[0] = maxbt;
			result[1] = dtable.maxDepth(bottomt, startpg);
			result[5] = 0;
			// find new spg if possible
			if(startpg != 'a') {
				if (dtable.maxBT('a', ddepth) > bottomt) {
					result[2] = dtable.minPG(bottomt, ddepth);
					result[3] = dtable.diveFPG((char) result[2], ddepth, bottomt);
					result[4] = dtable.ssTest((char) result[3], ddepth) ? 1 : 0;
				}
			}
		}
		else {
			fpg = dtable.diveFPG(startpg, ddepth, bottomt);
			safe = true;
			result[3] = fpg;
			ssrequired = dtable.ssTest(fpg, ddepth);
			result[4] = ssrequired ? 1 : 0;
		} 
		
	}
	
	public boolean isSafe() {
		return safe;
	}
	
	public int getBT() {
		return bottomt;
	}
	
	public int getDepth() {
		return ddepth;
	}
	
	public char getSPG() {
		return spg;
	}
	
	public char getFPG() {
		return fpg;
	}
	
	public char getMPG() {
		return dtable.minPG(bottomt, ddepth);
	}
	
	public boolean getSS() {
		return ssrequired;
	}
	
	public String getWarnings() {
		String string = new String();
		
		if(safe) string = "Safe Dive\n";
		else string = "Unsafe Dive, max bottom time at " + ddepth + "m" + " is " + result[0] + " minutes diving from " + ((spg == 'a') ? "the surface": "pressure group " + spg) + ".\n";
		if(diveid > 1 && spg == 'a') string = string + "Dive is unsafe because of previous dive\n";
		
		return string;
	}
	
	@Override
	public String toString() {
		String output =  "*********Dive " + diveid + "*********\n";
		
		output = output + "Bottom Time = " + bottomt + ", Depth = " + ddepth + " SPG = " + spg + ", FPG = " + fpg +"\n";
		output = output + this.getWarnings();
		if(safe) {
			if(ssrequired) {
				output = output + "SS Required\n";
			} else {
				output = output + "SS not required, but always recommended.\n";
			}
		}
		
		output = output + "*********************\n";
		return output;
	}
	
	public void draw() {
		
	}
}

File: client/src/components/DivePlan/dive.js
import DiveTable from './divetable.js'

export class Dive {
  constructor (id) {
    this.bottomt = 0
    this.ddepth = 0
    this.diveid = id
    this.spg = 'a'
    this.fpg = 'a'
    this.ssrequired = false
    this.safe = true
    this.result = null
  }

  updateDive (startpg, time, depth) {
    let maxbt = DiveTable.maxBT(startpg, depth)
    this.result = [time, depth, startpg, 'a', 0, 1]

    if (time === undefined) time = this.bottomt
    if (depth === undefined) depth = this.ddepth

    this.bottomt = time
    this.ddepth = depth
    this.spg = startpg
    this.result[2] = startpg

    // If time requested is past the maximum bottom time at depth & spg set return array to carry user recomendations for UI warnings
    if (maxbt < time || (this.diveid > 1 && startpg === 'a')) {
      this.safe = false
      this.result[0] = maxbt
      this.result[1] = DiveTable.maxDepth(time, startpg)
      this.result[5] = 0
      // find new spg if possible
      if (startpg !== 'a') {
        if (DiveTable.maxBT('a', depth) > time) {
          this.result[2] = DiveTable.minPG(time, depth)
          this.result[3] = DiveTable.diveFPG(this.result[2], depth, time)
          this.result[4] = DiveTable.ssTest(this.result[3], depth) ? 1 : 0
        }
      }
    } else {
      this.fpg = DiveTable.diveFPG(startpg, depth, time)
      this.result[3] = this.fpg
      this.ssrequired = DiveTable.ssTest(this.fpg, depth)
      this.result[4] = this.ssrequired ? 1 : 0
    }
  }

  isSafe () {
    return this.safe
  }

  getBT () {
    return this.bottomt
  }

  getDepth () {
    return this.ddepth
  }

  getSPG () {
    return this.spg
  }

  getFPG () {
    return this.fpg
  }

  getMPG () {
    return DiveTable.minPG(this.bottomt, this.ddepth)
  }

  getSS () {
    return this.ssrequired
  }

  getWarnings () {
    let string = null

    if (this.safe) string = 'Safe Dive\n'
    else string = 'Unsafe Dive, max bottom time at ' + this.ddepth + 'm' + ' is ' + this.result[0] + ' minutes diving from ' + ((this.spg === 'a') ? 'the surface' : 'pressure group ' + this.spg) + '.\n'
    if (this.diveid > 1 && this.spg === 'a') string = string + 'Dive is unsafe because of previous dive\n'

    return string
  }

  toString () {
    let output = '*********Dive ' + this.diveid + '*********\n'

    output = output + 'Bottom Time = ' + this.bottomt + ', Depth = ' + this.ddepth + ' SPG = ' + this.spg + ', FPG = ' + this.fpg + '\n'
    output = output + this.getWarnings()
    if (this.safe) {
      if (this.ssrequired) {
        output = output + 'SS Required\n'
      } else {
        output = output + 'SS not required, but always recommended.\n'
      }
    }

    output = output + '*********************\n'
    return output
  }
}

export default Dive
